### PR TITLE
Stanislav savenko

### PR DESCRIPTION
**Pull request by ssSavenko**
DONT MERGE
Onboarding branch

### CODE REVIEW

**Comment by dnillovna** [comment]
@ssSavenko name your PR according to our commit policy

**Comment by dnillovna** [comment]
@ssSavenko To avoid merge switch to PR to 'draft'

File: packages/stanislav-sequence/detectors.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * See also: https://datagrok.ai/help/develop/how-to/define-semantic-type-detectors
+ * The class name is comprised of <PackageName> and the `PackageDetectors` suffix.
+ * Follow this naming convention to ensure that your detectors are properly loaded.
+ */
+class StanislavSequencePackageDetectors extends DG.Package {
+    //tags: semTypeDetector
+    //input: column col
+    //output: string semType
+    detectNucleotides(col) {
+        console.log(col);
```
Comment (LaykDimon) [inline packages/stanislav-sequence/detectors.js:13]: unnecessary logs

File: packages/stanislav-sequence/detectors.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        if (col.type == "string" && DG.Detector.sampleCategories(col, (s) => {
```
Comment (LaykDimon) [inline packages/stanislav-sequence/detectors.js:14]: please make strict compares

File: packages/stanislav-sequence/detectors.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        }
+        return null;
+    }
+
+    //tags: semTypeDetector
+    //input: column col
+    //output: string semType
+    detectENAID(col) { 
+        console.log(col);
+        if (col.type == "string" && DG.Detector.sampleCategories(col, (s) => {
```
Comment (LaykDimon) [inline packages/stanislav-sequence/detectors.js:28]: same in here

File: packages/stanislav-sequence/src/Exercise2/CountSubsequenceJS.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+//name: CountSubsequenceJS 
+// input: string sequence
+// input: string subsequence
+// output: int count 
+//language: javascript
+
+export function complementWidget(sequence: any, subsequence: any): any {
```
Comment (LaykDimon) [inline packages/stanislav-sequence/src/Exercise2/CountSubsequenceJS.ts:7]: you can't use 'any' so much
additionallym, these are definitely the strings, why using any?

File: packages/stanislav-sequence/src/package.ts [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+
+//name: info
+export function info() {
+  grok.shell.info(_package.webRoot);
+}
+
+//name: complement
+//input: string nucleotides
+//output: string result
+export async function complement(nucleotide: string): Promise<string> {
```
Comment (LaykDimon) [inline packages/stanislav-sequence/src/package.ts:17]: why using async in here?

File: packages/stanislav-sequence/src/package.ts [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+  result = swapTwoWays(nucleotide, 'C', 'G');
+  return result;
+}
+
+//name: dna_nucleotide 
+//input: string nucleotides {semType: dna_nucleotide}
+//output: string result {semType: dna_nucleotide}
+export function dna_nucleotide(nucleotide: string): string {
+  // your code goes here
+  return nucleotide;
```
Comment (LaykDimon) [inline packages/stanislav-sequence/src/package.ts:28]: does this function has to look like this?

File: packages/stanislav-sequence/src/package.ts [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+}
+
+//name: complementWidget
+//tags: panel, widgets
+//input: string nucleotides {semType: dna_nucleotide}
+//output: widget result
+//condition: true
+export function complementWidget(nucleotide: any): any {
```
Comment (LaykDimon) [inline packages/stanislav-sequence/src/package.ts:36]: same for any here

File: packages/stanislav-sequence/src/package.ts [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+  return df;
+}
+
+//name: Add Tables
+//output: string t
+export async function addTables(): Promise<String> {
+  const files = await _package.files.list('', true);
+
+  const csvFiles = files.filter((f) => f.extension === 'csv');
+  let t = "";
```
Comment (LaykDimon) [inline packages/stanislav-sequence/src/package.ts:95]: single quotes

File: packages/stanislav-sequence/src/package.ts [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+  // Load every table and add a view for it
+  for (const file of csvFiles) {
+    const df = await _package.files.readCsv(file.fileName);
+    t = t + file + ' ';
+    grok.shell.addTableView(df);
+    // const df = await grok.data.loadTable(`${_package.webRoot}${file.path}`);
+    // const df = await grok.data.files.openTable(`System:AppData/${_package.name}/${file.fileName}`);
```
Comment (LaykDimon) [inline packages/stanislav-sequence/src/package.ts:102]: no deprecated comments needed

File: packages/stanislav-sequence/src/package.ts [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+  const col2 = df2.columns.byName(col.name).toList();
+  const col1Length = col1.length;
+
+  let resultDF = df1.append(df2);
+  resultDF.columns.addNew("Subsequence Count", "int")
+
+  let i = 0;
+  for (let item of resultDF.rows) {
+    if (i < col1Length) {
+      item["Subsequence Count"] = getCountOfSubsequencesInColumn(item[col.name], N, col2)
```
Comment (LaykDimon) [inline packages/stanislav-sequence/src/package.ts:124]: you could use just column.set(idx, value) for this, but this looks ok as well I guess

File: packages/stanislav-sequence/src/package.ts [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+      item["Subsequence Count"] = getCountOfSubsequencesInColumn(item[col.name], N, col1)
+    }
+    i++;
+  }
+
+  grok.shell.addTableView(resultDF);
+  return resultDF;
+}
+
+function getCountOfSubsequencesInColumn(subsequenceRowString: string, subseqSize: number, colValues: any): number {
```
Comment (LaykDimon) [inline packages/stanislav-sequence/src/package.ts:136]: what is colValues and why is it any?

File: packages/stanislav-sequence/src/package.ts [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+  let result = 0;
+  let subseqItems: Set<string> = new Set<string>();
+
+  for (let i = 0; i + subseqSize <= subsequenceRowString.length; i++) {
```
Comment (LaykDimon) [inline packages/stanislav-sequence/src/package.ts:140]: no braces needed here

File: packages/stanislav-sequence/src/package.ts [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+    subseqItems.add(subsequenceRowString.substring(i, i + subseqSize));
+  }
+
+  for (let item of colValues) {
```
Comment (LaykDimon) [inline packages/stanislav-sequence/src/package.ts:144]: same

File: packages/stanislav-sequence/src/package.ts [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+
+//name: nucleotideBoxCellRenderer
+//tags: cellRenderer
+//meta.cellType: dna_nucleotide
+//output: grid_cell_renderer result
+export function nucleotideBoxCellRenderer() {
+  return new NucleotideBoxCellRenderer();
+}
+
+export class NucleotideBoxCellRenderer extends DG.GridCellRenderer {
```
Comment (LaykDimon) [inline packages/stanislav-sequence/src/package.ts:160]: this should be in a separate file

File: packages/stanislav-sequence/src/package.ts [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+  get cellType() { return 'dna_nucleotide'; }
+
+
+  getDefaultSize(gridColumn: DG.GridColumn): { width?: number | null, height?: number | null } {
+    return { width: 200, height: 30 };
+  }
+
+  render(g: CanvasRenderingContext2D, x: number, y: number, w: number, h: number,
+    gridCell: DG.GridCell, cellStyle: DG.GridCellStyle) {
+    let seq = gridCell.cell.value;
```
Comment (LaykDimon) [inline packages/stanislav-sequence/src/package.ts:171]: switch let to const

File: packages/stanislav-sequence/src/package.ts [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+    let ctx = g.canvas.getContext('2d')!;
+    ctx.font = '11px courier';
+    // ...
+    let localx = x + 22
```
Comment (LaykDimon) [inline packages/stanislav-sequence/src/package.ts:175]: these magic values such as 22 and 11 should be in some constants

File: packages/stanislav-sequence/src/package.ts [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+
+  let df = DG.DataFrame.fromColumns([
+    DG.Column.fromList(DG.COLUMN_TYPE.STRING, 'ID', ["a", "b"]),
+    DG.Column.fromList(DG.COLUMN_TYPE.STRING, 'Sequence', ["c", "d"])
+  ]);
+
+  let inputData = 'coronavirus'
+  let grid = DG.Viewer.grid(df);
+
+  let limitInput = ui.intInput('How many rows: ', 100);
```
Comment (LaykDimon) [inline packages/stanislav-sequence/src/package.ts:231]: use new JS API - ui.input.int and ui.input.string

File: packages/stanislav-sequence/src/package.ts [hunk 18]
```diff
@@ -0,0 +1,333 @@
...
+  let queryInput = ui.stringInput('Query: ', inputData);
+ 
+  let button = ui.button('Preview', () => { 
+  
+    browseDataForENAID(queryInput.value, limitInput.value|| 10).then((result) =>{
```
Comment (LaykDimon) [inline packages/stanislav-sequence/src/package.ts:236]: missing space

File: packages/stanislav-sequence/src/package.ts [hunk 19]
```diff
@@ -0,0 +1,333 @@
...
+      browseDataForENAID(queryInput.value, limitInput.value|| 10).then((result) =>{
+        df = DG.DataFrame.fromColumns( [ DG.Column.fromStrings("EnaId", (result)) ] );  
+        grok.shell.addTableView(df);
+      });
+    })
+    .show();
+}
+
+function fetchEnaIdSequence(queryString: string, elementsCount: number): string {
+  return ""
```
Comment (LaykDimon) [inline packages/stanislav-sequence/src/package.ts:261]: what is it?

File: packages/stanislav-sequence/src/swap.ts [hunk 20]
```diff
@@ -0,0 +1,8 @@
+export async function swapTwoWays(stringToChange: string, characterToSwap1: string, characterToSwap2: string): Promise<string> {
+    let result = stringToChange.replaceAll(characterToSwap1, '\t' + characterToSwap1 + '\t')
+    result = result.replaceAll(characterToSwap2, '\t' + characterToSwap2 + '\t')
+    result = result.replaceAll('\t' + characterToSwap2 + '\t', characterToSwap1)
+    result = result.replaceAll('\t' + characterToSwap1 + '\t', characterToSwap2)
+
```
Comment (LaykDimon) [inline packages/stanislav-sequence/src/swap.ts:6]: why is there so many replaces?

File: packages/stanislav-sequence/src/tests/counter-test.ts [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+        const expectedValues = [0, 2, 1];
+        console.log(resultValue);
+        // const gainedValues = resultValue.getCol("N("+subsequence+")").getRawData();
+
+        if(resultValue.length !== expectedValues.length){
+            expect(1,2);
+        }
+
+        for (let i = 0; i < expectedValues.length; i++) {
+            expect(expectedValues[i],resultValue[i]);
```
Comment (LaykDimon) [inline packages/stanislav-sequence/src/tests/counter-test.ts:39]: you could just use function expectArray

**Pull request closed without merging**